File: Retos/Reto #16 - LA ESCALERA [Media]/ruby/test0n3.rb
# frozen_string_literal: true

def stairs_generator(input)
  stairs = downward_stairs(input.abs)

  stairs = upward_stairs(stairs) if input.positive?

  stairs.join("\n")
end

def downward_stairs(input)
  stairs = ['_']
  stairs[0] += '_' if input.zero?

  stair = 1
  while stair <= input
    stairs[stair] = "#{' ' * stairs[stair - 1].length}|_"
    stair += 1
  end
  stairs
end

def upward_stairs(negative_stairs)
  max_string_length = negative_stairs.last.length
  negative_stairs.map do |item|
    item += ' ' * (max_string_length - item.length)
    item.reverse
  end
end

# puts stairs_generator(-5)
# puts stairs_generator(4)

File: Retos/Reto #16 - LA ESCALERA [Media]/javascript/test0n3.js
const stairsGenerator = (input) => {
  let stairs = "";

  if (input <= 0) stairs = downwardStairs(Math.abs(input));

  if (input > 0) stairs = upwardStairs(input);

  return stairs.join("\n");
};

const downwardStairs = (input) => {
  let stairs = ["_"];
  if (input == 0) stairs[0] += "_";

  let stair = 1;
  while (stair <= input) {
    stairs.push(`${" ".repeat(2 * stair - 1)}|_`);
    stair += 1;
  }
  return stairs;
};

const upwardStairs = (input) => {
  let maxSpaceLength = 1 + 2 * input;
  let stairs = [`${" ".repeat(maxSpaceLength - 1)}_`];
  let stair = 1;
  while (stair <= input) {
    stairs.push(`${" ".repeat(maxSpaceLength - (2 * stair + 1))}_|`);
    stair += 1;
  }
  return stairs;
};

// console.log(stairsGenerator(0));
// console.log(stairsGenerator(-5));
// console.log(stairsGenerator(5));
